### ISSUE TITLE
Push SGD under 10ms on sparse parity

### ISSUE THREAD

**Issue by 0bserver07** [cybertronai/SutroYaro#4]
## Context

Yaroslav asked why SGD can't run in the same compute budget as LeCun's Spark 7 experiments. Current SGD takes 142ms (n=20/k=3, hidden=200). Every other method in the harness runs under 15ms. SGD is the outlier.

The "Spark 7 constraint": 1 second today = ~1 hour on 1980s hardware. 142ms = ~40 hours on Spark 7. Too long.

## Task

- Shrink hidden layer (16, 32, 64) and measure accuracy + time
- Reduce training samples (100, 200, 500 instead of 1000)
- Try early stopping (halt at 95% instead of 100%)
- Try curriculum (start small n, expand)
- Find the config that solves parity under 10ms with 100% accuracy

## Goal

SGD baseline that fits `bin/reproduce-all --budget 10`.

## Reason

Yaroslav's core principle: keep experiments under 1 second. SGD is the only method that violates this. If SGD can't get under 10ms, that's a finding too (SGD is fundamentally too slow for this problem at any config).

## How to start

```bash
PYTHONPATH=src python3 src/harness.py --method sgd --n_bits 20 --k_sparse 3 --hidden 32 --n_train 200 --json
```

## Files

- `src/harness.py` (read only, use existing SGD method)
- `research/search_space.yaml` (allowed parameter ranges)
- Log results to `findings/exp_sgd_speed.md`

**Comment by 0bserver07** [comment]
**Telegram context** (chat-yad, 2026-03-14):

> Yaroslav: "Sgd was used for spark 7. That's how Jan LeCun did his experiments."
> Yad: "question: do we want to test if SGD be pushed below 10ms on parity?"
> Yaroslav: "I think it makes sense to optimize things to be as efficient as they were for Yann LeCun, because modern systems have so much compute."
> Yaroslav: "My job on this project is to keep reminding people to keep experiments <1 second in duration."

**Update**: Ran the sweep. SGD floor is ~70ms (7 grokking epochs at ~10ms each). Can't hit 10ms. The phase transition is the bottleneck. Full findings in `findings/exp_sgd_speed.md`. This is a LOSS but a valid finding: the 10ms budget correctly filters SGD out for parity.

**Comment by 0bserver07** [comment]
Researched to conclusion: the answer is "don't push SGD further, use GF(2)."

**Evidence:** [`findings/exp_sgd_speed.md`](https://github.com/cybertronai/SutroYaro/blob/main/findings/exp_sgd_speed.md) — full hidden/LR/n_train sweep. Physical floor is ~70ms (7 grokking epochs × ~10ms). Also tried:

- L-BFGS — 35–60ms (close but unreliable)
- Sign SGD — 7.6ms on 3/5 seeds (below threshold but not robust)
- EGD — 12% worse than baseline

**Comparison:** GF(2) Gaussian elimination solves the same problem in 509µs — ~200× faster than the SGD floor. See DISCOVERIES.md "Current Best Methods" table.

**Conclusion:** the 10ms budget is correctly filtering SGD out in favor of algebraic methods. Parity is linear over GF(2) — the neural net was solving an easy problem the hard way.

Closing as answered.

---
_agent-0bserver07 (Claude Code) on behalf of Yad_
